doc/src/manual/embedding.md

```diff
@@ -284,6 +285,51 @@ args[1] = some_other_value;
 JL_GC_POP();
 ```
 
+Each scope must have only one call to `JL_GC_PUSH*`. Hence, if all variables cannot be pushed once by
+a single call to `JL_GC_PUSH*`, or if there is more than 6 variables to be pushed and using an array
+of arguments is not an option, then one can use inner blocks:
+
+```c
+jl_value_t *ret1 = jl_eval_string("sqrt(2.0)");
+JL_GC_PUSH1(&ret1);
+jl_value_t *ret2 = 0;
+{
+    jl_function_t *func = jl_get_function(jl_base_module, "exp");
+    ret2 = jl_call1(func, ret1);
+    JL_GC_PUSH1(&ret2);
+    // Do something with ret2.
+    JL_GC_POP();    // This pops ret2.
+}
+JL_GC_POP();    // This pops ret1.
+```
+
+If it is required to hold the pointer to a variable between functions (or block scopes), then it is
+not possible to use `JL_GC_PUSH*`. In this case, it is necessary to create and keep a reference to the
+variable in the Julia global scope. One simple way to accomplish this is to create a global `IdDict`
+that will hold the references, avoiding deallocation by the GC.
+
+```c
+// After initialization, create a global `IdDict` called `refs`.
+jl_eval_string("refs = IdDict()");
+
+...
+
+// `var` is the variable we want to protect between function calls.
+jl_value_t* var = 0;
+
+...
+
+var = jl_eval_string("sqrt(2.0)");
+
+// To protect `var`, add its reference to `refs`.
+jl_value_t* refs = jl_eval_string("refs");
+jl_function_t* setindex = jl_get_function(jl_base_module, "setindex!");
+jl_call3(setindex, refs, var, var);
```

Comment:
Actually it's worse than that. This code does not guarantee that `var` is kept alive. There's no guarantee that the pointer stored in `refs` is `var`, it could be a re-box. This is true for any immutables. If you really want to keep a pointer you hold, you have to avoid going through julia code at all. A `Vector{Any}` can be operated easily in C.

Reply:
Hi @yuyichao,

I am not sure if I understood. Do you mean that I need to make every variable a `Vector{Any}` and push it to `refs`?

Reply:
You don't need to push the `Vector{Any}` to anything. That can be your container. You can use the index to delete it later if your usage pattern is simple this should be the easiest way.

This is only an issue with immutables, mutable types doesn't have this problem (i.e. they are guaranteed to round trip through julia code with the same pointer).

If your usage pattern is complex, you can either just copy the implementation of `IdDict` or if you don't mind the extra allocation, you can wrap the pointer in a `RefValue{Any}` or any other mutable type with equivalent layout and put that in the `IdDict`.

Reply:
> Why not just add var to a global variable like:

Well, if you only have a single/fixed number of variable(s) to handle, there are many other ways to handle this and this doesn't seem like the best way either.... (e.g. unless you also only need this once it's much better to just use a global container rooted as a global or local variable and just operate on that container without round trip through the julia runtime.)

Reply:
Nice! I will try to change the PR to reflect that!